llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```
@@ -1018,6 +1024,25 @@ bool SPIRVInstructionSelector::selectLoad(Register ResVReg,
                                           MachineInstr &I) const {
   unsigned OpOffset = isa<GIntrinsic>(I) ? 1 : 0;
   Register Ptr = I.getOperand(1 + OpOffset).getReg();
+
+  auto *PtrDef = getVRegDef(*MRI, Ptr);
+  auto *IntPtrDef = dyn_cast<GIntrinsic>(PtrDef);
+  if (IntPtrDef &&
+      IntPtrDef->getIntrinsicID() == Intrinsic::spv_resource_getpointer) {
```

So if the codegen/opti doesn't keep the `getpointer`/`load` duo, but adds another temporary in between, this will fail silently and generate an OpLoad no?
Could reg2mem be an issue?

---

Yes, we are relying on pattern matching, which can be easily broken. If optimizations are not run, then we will not be able to do this.

The HLSL codegen should not have a temp between the intrinsic and the load/store. There may be a temporary between the `handlefrombinding` call and the `getpointer`. If there is, then there is a store of an opaque type, which is not allowed in spir-v.

> Could reg2mem be an issue?

That pass only more values that are live across basic block boundaries, right? If so, I would not expect to create a temp between the `getpointer` and the load/store. They will be generated side-by-side.

It could add a temp between `handlefrombinding` and the `getpointer`. However, a simple call to `mem2reg` should fix that. Don't we already do that after the structurizer?

---

You are write that if the optimizer commons the `getpointer` call, and more it to another basic block, then `reg2mem` could add a temp. Again, we will need `mem2reg`.

---

From what I've seen, reg2mem is not that clever. Seems like every single register is dumbly stored into a variable, and loaded just before use.

We do have a mem2reg pass after the structurizer (PromoteMemoryToRegisterPass) so that should be OK.

---

Looks like the getpointer calls are not merged into one, so looks OK!